File: js/test_money.js
const assert = require('assert');
const Money = require('./money');
const Bank = require('./bank');
const Portfolio = require('./portfolio');

class MoneyTest {
    setup() {
        this.bank = new Bank();
        this.bank.addExchangeRate("EUR", "USD", 1.2)
        this.bank.addExchangeRate("USD", "KRW", 1100)
    }

    testMultiplication() {
        let tenEuro = new Money(10, "EUR");
        let twentyEuro = new Money(20, "EUR")
        assert.deepStrictEqual(tenEuro.times(2), twentyEuro);
    }

    testDivision() {
        let originalMoney = new Money(4002, "KRW");
        let actualMoneyAfterDivision = originalMoney.divide(4);
        let expectedMoneyAfterDivision = new Money(1000.5, "KRW");
        assert.deepStrictEqual(actualMoneyAfterDivision, expectedMoneyAfterDivision);
    }

    testAddition() {
        let fiveDollar = new Money(5, "USD");
        let tenDollar = new Money(10, "USD");
        let fifteenDollar = new Money(15, "USD");
        let portfolio = new Portfolio();
        portfolio.add(fiveDollar, tenDollar);
        assert.deepStrictEqual(portfolio.evaluate(new Bank(), "USD"), fifteenDollar)
    }

    testAdditionOfDollarsAndEuros() {
        let fiveDollars = new Money(5, "USD");
        let tenEuros = new Money(10, "EUR");
        let portfolio = new Portfolio();
        portfolio.add(fiveDollars, tenEuros)
        let expectedValue = new Money(17, "USD");
        assert.deepStrictEqual(portfolio.evaluate(this.bank, "USD"), expectedValue)
    }

    testAdditionOfDollarsAndWon() {
        let oneDollar = new Money(1, "USD");
        let elevenHundredWon = new Money(1100, "KRW");
        let portfolio = new Portfolio();
        portfolio.add(oneDollar, elevenHundredWon)
        let expectedValue = new Money(2200, "KRW");
        assert.deepStrictEqual(portfolio.evaluate(this.bank, "KRW"), expectedValue)
    }

    testAdditionWithMultipleMissingExchangeRates() {
        let oneDollar = new Money(1, "USD")
        let oneEuro = new Money(1, "EUR")
        let oneWon = new Money(1, "KRW")
        let portfolio = new Portfolio()
        portfolio.add(oneDollar, oneEuro, oneWon)
        let expectedError = new Error("Missing exchange rate(s):[USD->Kalganid,EUR->Kalganid,KRW->Kalganid]")
        assert.throws(() => portfolio.evaluate(this.bank, "Kalganid"), expectedError)
    }

    testConversionWithDifferentRatesBetweenTwoCurrencies() {
        let tenEuros = new Money(10, "EUR")
        assert.deepStrictEqual(this.bank.convert(tenEuros, "USD"), new Money(12, "USD"))
        this.bank.addExchangeRate("EUR", "USD", 1.3)
        assert.deepStrictEqual(this.bank.convert(tenEuros, "USD"), new Money(13, "USD"))
    }

    testConversionWithMissingExchangeRates() {
        let bank = new Bank();
        let tenEuros = new Money(10, "EUR");
        let expectedError = new Error("EUR->Kalganid")
        assert.throws(function () {
            bank.convert(tenEuros, "Kalganid")
        }, expectedError)
    }

    runAllTests() {
        let testMethods = this.getAllTestMethods()
        this.randomizeTestOrder(testMethods).forEach(m => {
            console.log("Running: %s()", m)
            let method = Reflect.get(this, m);
            try {
                this.setup()
                Reflect.apply(method, this, []);
            } catch (e) {
                if (e instanceof assert.AssertionError) {
                    console.log(e)
                } else {
                    throw e
                }
            }
        })
    }

    randomizeTestOrder(testMethods) {
        for (let i = testMethods.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [testMethods[i], testMethods[j]] = [testMethods[j], testMethods[i]]
        }
        return testMethods
    }

    getAllTestMethods() {
        let moneyPrototype = MoneyTest.prototype
        let allProps = Object.getOwnPropertyNames(moneyPrototype);
        let testMethods = allProps.filter(p => {
            return typeof moneyPrototype[p] === 'function' && p.startsWith('test')
        })
        return testMethods
    }
}

new MoneyTest().runAllTests();

File: py/test_money.py
import unittest

from money import Money
from portfolio import Portfolio
from bank import Bank


class TestMoney(unittest.TestCase):

    def setUp(self):
        self.bank = Bank()
        self.bank.addExchangeRate("EUR", "USD", 1.2)
        self.bank.addExchangeRate("USD", "KRW", 1100)

    def testMultiplication(self):
        tenEuros = Money(10, "EUR")
        twentyEuros = Money(20, "EUR")
        self.assertEqual(twentyEuros, tenEuros.times(2))

    def testDivision(self):
        originalMoney = Money(4002, "KRW")
        expectedMoneyAfterDivision = Money(1000.5, "KRW")
        self.assertEqual(expectedMoneyAfterDivision, originalMoney.divide(4))

    def testAddition(self):
        fiveDollars = Money(5, "USD")
        tenDollars = Money(10, "USD")
        fifteenDollars = Money(15, "USD")
        portfolio = Portfolio()
        portfolio.add(fiveDollars, tenDollars)
        self.assertEqual(fifteenDollars, portfolio.evaluate(self.bank, "USD"))

    def testAdditionDollarsAndEuros(self):
        fiveDollars = Money(5, "USD")
        tenEuros = Money(10, "EUR")
        expectedValue = Money(17, "USD")
        portfolio = Portfolio()
        portfolio.add(fiveDollars, tenEuros)
        actualValue = portfolio.evaluate(self.bank, "USD")
        self.assertEqual(expectedValue, actualValue, "%s != %s" % (expectedValue, actualValue))

    def testAdditionDollarsAndWon(self):
        oneDollar = Money(1, "USD")
        elevenHundretWon = Money(1100, "KRW")
        portfolio = Portfolio()
        portfolio.add(oneDollar, elevenHundretWon)
        expectedValue = Money(2200, "KRW")
        actualValue = portfolio.evaluate(self.bank, "KRW")
        self.assertEqual(expectedValue, actualValue, "%s != %s" % (expectedValue, actualValue))

    def testAdditionWithMultipleMissingExchangeRates(self):
        oneDollar = Money(1, "USD")
        oneEuro = Money(1, "EUR")
        oneWon = Money(1, "KRW")
        portfolio = Portfolio()
        portfolio.add(oneDollar, oneEuro, oneWon)
        with self.assertRaisesRegex(
                Exception, "Missing exchange rate\(s\):\[USD->Kalganid,EUR->Kalganid,KRW->Kalganid\]",
        ):
            portfolio.evaluate(self.bank, "Kalganid")

    def testConversion(self):
        bank = Bank()
        bank.addExchangeRate("EUR", "USD", 1.2)
        tenEuros = Money(10, "EUR")
        actual = bank.convert(tenEuros, "USD")
        expected = Money(12, "USD")
        self.assertEqual(actual, expected, "%s != %s" % (expected, actual))

    def testConversionWithMissingExchangeRate(self):
        bank = Bank()
        tenEuros = Money(10, "EUR")
        with self.assertRaisesRegex(Exception, "EUR->Kalganid"):
            bank.convert(tenEuros, "Kalganid")


if __name__ == "__main__":
    unittest.main()
